File: packages/embed-js/src/embed.ts
/**
 * High-level embed() API: markdown in, EmbeddingsPayload out, ready to drop
 * into pack({ embeddings: ... }).
 */

import type { EmbeddingChunk, EmbeddingSpace, EmbeddingsPayload } from '@cvfile/sdk';
import { chunkMarkdown, type ChunkingMode } from './chunk.js';
import { createTransformersBackend } from './transformers-backend.js';
import { DEFAULT_MODEL, type EmbeddingBackend } from './types.js';

export interface EmbedOptions {
  /** HF model id; defaults to BGE-M3 (Xenova/bge-m3). */
  model?: string;
  /** Pinned model revision; recorded in the payload for reproducibility. */
  modelRevision?: string;
  /** Chunking strategy; default 'section'. */
  chunking?: ChunkingMode;
  /** Bring-your-own backend (e.g. an OpenAI/Voyage adapter). */
  backend?: EmbeddingBackend;
}

export async function embed(markdown: string, opts: EmbedOptions = {}): Promise<EmbeddingsPayload> {
  const chunks = chunkMarkdown(markdown, { mode: opts.chunking ?? 'section' });
  const backend =
    opts.backend ??
    createTransformersBackend(
      opts.modelRevision !== undefined
        ? { model: opts.model ?? DEFAULT_MODEL, modelRevision: opts.modelRevision }
        : { model: opts.model ?? DEFAULT_MODEL },
    );

  const matrix = await backend.embed(chunks.map((c) => c.text));
  if (matrix.vectors.length !== chunks.length) {
    throw new Error(`Backend returned ${matrix.vectors.length} vectors for ${chunks.length} chunks`);
  }

  const embeddingChunks: EmbeddingChunk[] = chunks.map((c, i) => ({
    id: c.id,
    textOffset: c.textOffset,
    textLength: c.textLength,
    vector: matrix.vectors[i]!,
  }));

  const space: EmbeddingSpace = {
    model: backend.model,
    modelRevision: backend.modelRevision,
    dimension: matrix.dimension,
    metric: backend.metric,
    normalized: backend.normalized,
    chunking: opts.chunking ?? 'section',
    chunks: embeddingChunks,
  };

  return { formatVersion: 1, spaces: [space] };
}

File: sdks/python/src/cvfile/embed/_embed.py
"""High-level embed() API: markdown → EmbeddingsPayload."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from cvfile.embed._chunk import ChunkingMode, chunk_markdown
from cvfile.embed._embeddings import (
    EmbeddingChunk,
    EmbeddingSpace,
    EmbeddingsPayload,
)
from cvfile.embed._huggingface import HuggingFaceBackend


class EmbeddingBackend(Protocol):
    model: str
    model_revision: str
    metric: str
    normalized: bool

    def embed(self, texts: list[str]) -> tuple[list[tuple[float, ...]], int]: ...


@dataclass(slots=True)
class EmbedOptions:
    model: str = "BAAI/bge-m3"
    model_revision: str | None = None
    chunking: ChunkingMode = "section"
    backend: EmbeddingBackend | None = None


def embed(markdown: str, options: EmbedOptions | None = None) -> EmbeddingsPayload:
    opts = options or EmbedOptions()
    chunks = chunk_markdown(markdown, mode=opts.chunking)

    backend: EmbeddingBackend = opts.backend or HuggingFaceBackend(
        model=opts.model,
        model_revision=opts.model_revision or "main",
    )

    vectors, dimension = backend.embed([c.text for c in chunks])
    if len(vectors) != len(chunks):
        raise RuntimeError(f"Backend returned {len(vectors)} vectors for {len(chunks)} chunks")

    embedding_chunks = tuple(
        EmbeddingChunk(
            id=chunks[i].id,
            text_offset=chunks[i].text_offset,
            text_length=chunks[i].text_length,
            vector=vectors[i],
        )
        for i in range(len(chunks))
    )

    space = EmbeddingSpace(
        model=backend.model,
        model_revision=backend.model_revision,
        dimension=dimension,
        metric=backend.metric,  # type: ignore[arg-type]
        normalized=backend.normalized,
        chunking=opts.chunking,
        chunks=embedding_chunks,
    )
    return EmbeddingsPayload(format_version=1, spaces=(space,))
